File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0946_Validate_Stack_Sequences.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Stack;

public class Question_0946_Validate_Stack_Sequences {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> stack = new Stack();
        int j = 0;
        for (int x : pushed) {
            stack.push(x);
            while (!stack.isEmpty() && j < popped.length && stack.peek() == popped[j]) {
                stack.pop();
                j++;
            }
        }
        return j == popped.length;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0946_Validate_Stack_Sequences.cc
#include <vector>
#include <stack>

using namespace std;

class Solution {
 public:
  bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
    stack<int> st;
    int n = pushed.size(), i = 0;
    for (int j = 0; j < popped.size(); j++) {
      while (i < pushed.size() && (st.empty() || st.top() != popped[j])) {
        st.push(pushed[i++]);
      }
      if (st.top() != popped[j]) {
        break;
      }
      st.pop();
    }
    return st.empty();
  }
};
